Custom Exception Handlers at toplevel

AndrewLitteken opened the pull request:
This is in response to issue [#7156](https://github.com/ocaml/ocaml/issues/7156) where the OCaml Top-level level does not use any custom printers if they are registered.

There is a new function in `Printexc.ml` called `to_string_opt` that returns an option that is None where there are no printers registers, and Some string when there are printers that have been registered to resolve the Exception.

The addition involved adding to the `print_out_exception` function found in `types/oprint.ml`.  For all Exceptions excluding Interrupts and Out of Memory errors, the result of the new function `Printexc.to_string_opt` is called against the exception.  If None is returned, the previous method is used, if Some is returned then the string generated from the printers is used.

Closes [#7156](https://github.com/ocaml/ocaml/issues/7156)

AndrewLitteken committed:
if printers exist ocaml toplevel will use them

AndrewLitteken committed:
adding to changelog

gasche commented on Changes:
```diff
@@ -161,6 +161,9 @@ Working version
 
 ### Bug fixes:
 
+- #7156: Create Mechanism so that OCaml Top Level uses custom printers 
```
lowercase is more standard, and "exception" should be in the description: "make the toplevel use custom exception printers if they are available>

dra27 replied:
Trailing space on this line (see details of the hook in [CONTRIBUTING.md](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#coding-guidelines))

dra27 commented on stdlib/printexc.ml:
```diff
@@ -69,6 +69,11 @@ let to_string x =
 
+let to_string_opt x =
+  match !printers with
+      [] -> None
```
If this stands, please indent as in other `match` cases (so reduce indentation) and include a `|` on the first case.

dra27 commented on stdlib/printexc.mli:
```diff
@@ -22,6 +22,11 @@ val to_string: exn -> string
+val to_string_opt: exn -> string option
+(** [Printexc.to_string_opt e] returns a string option containing None if
+  there are no registered printers and Some s for a representation if there
+  are registered printers. *)
```
Again, if it stands only: this could be more terse, in keeping with other descriptions. The type tells us that the function returns `string option` so it doesn't need stating in the description. Something more like `[Printexc.to_string_opt e] returns [None] if there are no registered printers and [Some (to_string e)] otherwise.`

trefis replied:
Nitpick, but I'd make the name more explicit, something like: `to_pretty_string_opt`.
(feel free to ignore this comment)

gasche reviewed (commented):
Thanks for your contribution! This goes in the right direction but it's not satisfying yet. I think that the original idea of the PR was to *decompose* the implementation of `to_string` into a more modular `to_string_opt` function and a "default printing behavior" which is called when none of the registered printers work. In particular, if the registered printers all return `None` on an exception, then `to_string_opt` should also return `None`. This is not what your current implementation does.

I think you should try to split the `to_string` implementation into `to_string_opt` and `to_string_default` (you could even expose this latter function to the user as well, as it may be useful).

I would also be interested in seeing tests for this feature/bug, as an expect-style testfile named `testsuite/tests/tool-toplevel/uncaught_exceptions` (for an example of how to write expect-style tests, see for example `testsuite/tests/typing-misc/variant.ml`. In particular, there should be a test for the case where there are some registered printers, but they all return `None` on the exception raised.

nojb commented on stdlib/printexc.ml:
```diff
@@ -69,6 +69,11 @@ let to_string x =
+  match !printers with
+      [] -> None
+    | _ -> Some (to_string x)
+
```
This is not the right semantics: `!printers <> []` does not imply that there is a printer for `x` in `!printers`. Rather, I suggest:

1) Rename `to_string` into `to_string_opt` and change the code so that it returns an option with `None` in the case where there is no registered printer (right now it uses `Obj.magic` to produce a representation in this case).

2)  Re-implement `to_string` in terms of `to_string_opt` (the previous `Obj.magic` code would be moved here to handle the `None` case).

dra27 commented:
Hah - @gasche has hit submit on his review fractionally before me! I would add only that I don't think the function should be `to_string_opt` - let's keep `foo` and `foo_opt` meaning that `foo_opt` returns `None` when `foo` would raise.

I wonder if, for example, what's wanted is something like `fold_printers`?

gasche commented:
Yes, I agree `to_string_opt` is not the best name; `fold_printers` is nice, or even just `use_printers` may be fine. Also, in the `.mli` file this function should go right after `register_printer`, I think, as it assumes the knowledge of the details of the registration mechanism.

dra27 commented:
(The CI failure for 32-bit was of course #8585, which I've just merged)

Armael commented:
(This is great! I wanted to fix that bug a long time ago, but couldn't find how at that time.)

AndrewLitteken committed:
Merge pull request #1 from ocaml/trunk

updating to trunk of ocaml

AndrewLitteken committed:
updating changelog to be more concise

AndrewLitteken committed:
breaking to_string up and creating Printexc.use_printers

This changes Printexc.to_string into a match on Printexc.use_printers which
returns None if no printer exists or matches. Then a default using the
previous to_string is run.  Both of these new functions are surfaced to
the user.

The top level attempts to use use_printers, and if None is returned, uses
the previous exception printing.

AndrewLitteken committed:
Merge branch 'trunk' into custom-exceptions

AndrewLitteken committed:
adjusting print on oprint

AndrewLitteken commented:
I have made the restructuring changes that were suggested and renamed `to_string_opt` to `use_printers`.

I'm having some trouble creating expect tests for these since whenever an exception is raised computation ends.  Do you have any suggestions to work around this?

AndrewLitteken committed:
fixing typo error

gasche commented:
The following source file works on my machine as `testsuite/tests/tool-toplevel/uncaught_exceptions.ml`:

```ocaml
(* TEST
   * expect
*)

raise Exit;;
[%%expect {|
Exception: Stdlib.Exit.
|}];;

exception Foo of string;;
[%%expect {|
exception Foo of string
|}];;

raise (Foo "bar");;
[%%expect {|
Exception: Foo "bar".
|}];;
```

The workflow I use to create this file is to start with empty expect blocks:

```ocaml
[%%expect {|
|}];;
```

after each toplevel phrase, and then use (from the directory `tool-toplevel/`) the command `../../../ocamltest/ocamltest.opt -promote uncaught_exceptions.ml` to populate it with expected outputs.

Armael commented on typing/oprint.ml:
Nitpick: the `|` should be aligned with `match`

nojb commented on Changes:
```diff
@@ -167,6 +167,8 @@ Working version
 
 ### Bug fixes:
 
+- #7156: make top level use custom printers if they are available
```
Please also add the PR number.

nojb commented on stdlib/printexc.mli:
```diff
@@ -22,6 +22,10 @@ val to_string: exn -> string
 
+val to_string_default: exn -> string
+(** [Printexc.to_string_opt e] returns a string that would be generated
+    without registered printers of exception [e] *)
```
- `s/to_string_opt/to_string_default`
- wording suggestion: `[Printexc.to_string_default e] returns a string representation of the exception [e], ignoring all registered exception printers.`.

dra27 replied:
It also needs `@since 4.09`

dra27 commented on Changes:
```diff
@@ -167,6 +167,8 @@ Working version
 ### Bug fixes:
 
+- #7156: make top level use custom printers if they are available
+  (Andrew Litteken, report by Martin Jambon, review by )
```
And don't forget to add the (various!) reviewers

dra27 commented on stdlib/printexc.ml:
```diff
@@ -43,32 +43,39 @@ let fields x =
         (match try hd x with _ -> None with
-        | Some s -> s
+        | Some s -> Some s
         | None -> conv tl)
```
Stylistically, this could more neatly be:

```ocaml
let result = (try hd x with _ -> None) in
if result <> None then result else conv tl
```
or
```ocaml
  (try
     hd x
   with _ -> conv tl)
```

gasche replied:
(I'm not asking for a change to the PR because I think the current version is okay but) personally I found the earlier version more readable: it's easier for me to read a pattern-matching on an option type and detect the success and failure case than to understand what `if result <> None` is doing.

A third option would be to use the `match .. with exception` feature to avoid the intermediate `try .. with`.

```ocaml
match hd x with
| None | exception _ -> conv tl
| Some s -> Some s
```

if one is disturbed by the lack of sharing in the last case (I'm not), it's also possible to write it as

```ocaml
| (Some _) as result -> result
```

AndrewLitteken replied:
Since I had to fix documentation, I went ahead and went with the third option, I agree it is more readable

dra27 commented on stdlib/printexc.mli:
```diff
@@ -95,6 +99,10 @@ val register_printer: (exn -> string option) -> unit
 
+val use_printers: exn -> string option
+(** [Printexc.use_printers e] returns [None] if there are no registered
+    printers and [Some (to_string e)] otherwise.*)
```
This needs `@since 4.09`

dra27 commented on stdlib/printexc.ml:
A refactoring nit: but this could now be

```ocaml
let to_string_default = function
```
...

dra27 commented on stdlib/printexc.ml:
... with `x` here instead

dra27 reviewed (commented):
The API changes look fine, thanks. Some stylistic suggestions, and some documentation changes.

AndrewLitteken committed:
updating changes

AndrewLitteken committed:
adding uncuaght excpetion tests

AndrewLitteken committed:
formatting and documentation changes

AndrewLitteken committed:
Merge pull request #2 from ocaml/trunk

merging trunk

AndrewLitteken committed:
Merge branch 'trunk' into custom-exceptions

AndrewLitteken committed:
fixing white space issues

AndrewLitteken commented:
I believe all the style and formatting changes are present now. Are there anymore tests that should be added?

gasche commented on testsuite/tests/tool-toplevel/uncaught_exceptions.ml:
```diff
@@ -0,0 +1,39 @@
+raise Not_found;;
+[%%expect{|
+- : unit = ()
+- : unit = ()
```
It would be more readable if each sentence had its own expect test, instead of this "interesting" sentence having an expect-block that starts with two unrelated outputs (I thought this was an error at first).

gasche reviewed (approved):
Thanks! This looks good to me -- see a minor change suggestion inline. I won't merge right ahead to let my esteemed reviewer colleagues comment if they wish to. (Feel free to ping me if nothing has happened by the end of the week.)

There remains a minor glitch, which is that `Printexc.to_string` will in some case print a dedicated message coming from `to_string_default` by catching a few built-in exceptions (for example `Assert_failure`), which neither the compilers nor the toplevel seem to handle correctly in my test (the toplevel has some builtin-in logic of its own, which shows slightly different messages for a *subset* of those built-in-exceptions). It would have been nice to unify this a bit more, but I'm not sure what the right choice is there, it is not directly related to the problem reported in the PR, I don't want to delay the nice work and ultimately I think it's not a very important problem.

AndrewLitteken committed:
making test uncaught_exceptions.ml more readable

AndrewLitteken commented:
@gasche I think I updated the test as you suggested to be more readable.  Just pinging you to merge the pull request.

dra27 commented on stdlib/printexc.mli:
```diff
@@ -95,6 +101,12 @@ val register_printer: (exn -> string option) -> unit
 
+val use_printers: exn -> string option
+(** [Printexc.use_printers e] returns [None] if there are no registered
+    printers and [Some (to_string e)] otherwise.
```
This comment needs completely updating

AndrewLitteken replied:
I think this was referring to the `[Some (to_string e)]` it has been updated for says `[Some s] where s is the resulting string`

dra27 reviewed (commented):
There’s a doc comment which has got out of sync with all the changes, but otherwise LGTM

AndrewLitteken committed:
fixing documentation, readibility of use_printers

gasche commented:
I went ahead and merged. Thanks @AndrewLitteken and all reviewers!
